HPC - 8242 Stable version of FTSAdminRewrite without 'Add flow' functionality

Onitoxan opened the pull request:
# Description
FTS Admin Rewrite (FTS Admin v2) first pull request for Flow table, Pending Flow Table, Organization Table and CRUD page, Keyword page and CRUD functionalities.
# Features

- Feature Parity (Merge Organization and Merge Keywords functionalities)
- Support for TailwindCSS
- New custom form fields added in `hpc-ui`
- 🌐 English, Spanish, French, Simplified Chinese and Arabic support
- [New Feature] Reordering of columns on tables
- [New Feature] Download XLSX files of current filters

And a few more, for further readability please follow to the commit history.

# Next steps

- The content of this PR doesn't include Add / Edit Flow functionality (Matyas part)
- Test and open new tickets with parts of the application that are not properly working or is not doing good feature parity

Onitoxan committed:
Update package.json and TailwindCSS config files

Onitoxan committed:
Update libs/hpc-data hpc/live and hpc/dummy

Onitoxan committed:
Update libs/hpc-core

Onitoxan committed:
Update libs/hpc-ui

Onitoxan committed:
Add UseQueryParams support

Onitoxan committed:
Add paths

Onitoxan committed:
Update i18n

Onitoxan committed:
Add components

Onitoxan committed:
Add pages

Pl217 commented on apps/hpc-ftsadmin/src/app/components/filters/filter-pending-flows-table.tsx:
```diff
@@ -0,0 +1,281 @@
+  const { lang, env } = useContext(AppContext);
+  const environment = env();
+
+  const FORM_VALIDATION = object().shape({
+    flowStatus: string(),
+    dataProvider: string(),
+    reporterRefCode: number()
+      .positive()
+      .typeError('Only positive integers are accepted'),
+    sourceOrganizations: array().of(
+      object().shape({ label: string(), id: string() })
+    ),
+    sourceCountries: array().of(
+      object().shape({ label: string(), id: string() })
+    ),
+    destinationOrganizations: array().of(
+      object().shape({ label: string(), id: string() })
+    ),
+    destinationCountries: array().of(
+      object().shape({ label: string(), id: string() })
+    ),
+    destinationUsageYears: array().of(
+      object().shape({ label: string(), id: string() })
+    ),
+  });
```
By reading [official Formik documentation on validation](https://formik.org/docs/guides/validation), they have "baked-in support for schema-based form-level validation through Yup". Validation in Formik comes at form level and field level.

For form level validation, you can provide Yup codec via [`<Formik validationSchema>`](https://formik.org/docs/guides/validation#validationschema), but you can also use [`<Formik validate>`](https://formik.org/docs/guides/validation#validate) which accepts a method with `values` argument.

Since we're extensively using `io-ts` for runtime type checking, we can use it define codecs for usage inside `validate` method for form validation. You'll need to define custom codecs for things like positive integer, but there are examples of doing so in the codebase.

Pl217 commented on libs/hpc-live/src/lib/model.ts:
```diff
@@ -394,6 +407,332 @@ export class LiveModel implements Model {
+       * TODO: Dynamically fetch only necessary fields, Ex: if we don't display 'NewMoney' we shouldn't ask for it
+       */
+      searchFlows: (params) => {
+        const query = `query {
+          searchFlows${searchFlowsParams(params)} {
+            total
+            flows {
+              id
+              updatedAt
+              amountUSD
```
Can you explore using some GraphQL client library like Apollo client?

Onitoxan replied:
I was looking into it and it could be implemented. The biggest advantage of using Apollo client (or any other), is the automatic management of the cache. In terms of code, the syntax is pretty similar actually, is just using a template string preceded by `gql`. If you want we could create a ticket to add Apollo client :)

Pl217 commented on package.json:
```diff
@@ -106,8 +112,26 @@
+    "react-number-format": "^5.3.1",
+    "twin.macro": "^3.3.1",
+    "xlsx": "^0.18.5"
+  },
```
Before your changes, we had all dependencies under `devDependencies`. I thought that is some policy of Nx, but I cannot find anything saying that in their docs. Documentation page about [Nx console command](https://nx.dev/recipes/nx-console/console-add-dependency-command) to add a dependency has separate commands for prod and dev dependencies. Their [dependency management page](https://nx.dev/concepts/more-concepts/dependency-management) only deals with strategies.

Thus, you can abandon the approach of putting everything in `devDependencies` if you want to, but do it consistently. For example, `use-query-params` isn't a dev dependency, but you put it there. If you want to keep the current approach, the easiest thing would be to move all of them under `devDependencies` and leave to Nx to figure out which application depends on which libraries.

Pl217 commented on libs/hpc-dummy/src/lib/dummy.ts:
A bunch of `params` arguments are unused in this file

Pl217 commented on libs/hpc-data/src/lib/categories.ts:
It hurts me deeply that you needed to write this :( We've got model definitions in multiple places and I wish to have one "source of truth", where all type definitions would be. For example, I wish we had a common place where `category` model would be defined. We [have it in `hpc-api-core`](https://github.com/UN-OCHA/hpc-api-core/blob/33756b1c87fd73cf3910cafe81536acba9a5abcb/src/db/models/category.ts), but that is library shared **between APIs**, not meant to be used in frontends.

Even if we had a common type definitions, using them would require some refactoring to avoid the friction, where we would need to access each property individually. But, we need to resolve this. Imagine a new category group being added in the future. Would you remember to update this file, but also update in [hpc-api-core](https://github.com/UN-OCHA/hpc-api-core/blob/33756b1c87fd73cf3910cafe81536acba9a5abcb/src/db/models/categoryGroup.ts#L5-L32) and potentially other places? I would definitely forget.

I am not saying you did anything wrong, just a little rant because we should have done this before and have it ready for you.

Pl217 commented on libs/hpc-ui/src/lib/components/form-fields/async-autocomplete-field.tsx:
```diff
@@ -0,0 +1,247 @@
+  const [options, setOptions] = useState<Array<FormObjectValue>>([]);
+  const [data, setData] = useState<Array<FormObjectValue>>([]);
+  const [isFetch, setIsFetch] = useState(false);
+  const [isUsageYear, setisUsageYear] = useState(false);
```
Camel case
```suggestion
  const [isUsageYear, setIsUsageYear] = useState(false);
```

Pl217 commented on libs/hpc-ui/src/lib/components/form-fields/async-autocomplete-field.tsx:
```diff
@@ -0,0 +1,247 @@
+    result: APIAutocompleteResult
+  ): result is usageYears.GetUsageYearsResult {
+    return usageYears.GET_USAGE_YEARS_RESULT.is(result);
+  }
```
I am so happy that you learned about type guards in Typescript

Pl217 commented on libs/hpc-ui/src/lib/components/form-fields/async-autocomplete-field.tsx:
```diff
@@ -0,0 +1,247 @@
+        const parsedResponse = response.map((responseValue) => {
+          if (isUsageYearsResult(response)) {
+            return {
+              displayLabel: (responseValue as usageYears.UsageYear).year,
```
You should avoid using the type assertion like this, it can be really dangerous

Pl217 commented on libs/hpc-ui/src/lib/components/form-fields/async-autocomplete-field.tsx:
```diff
@@ -0,0 +1,247 @@
+    result: APIAutocompleteResult
+  ): result is organizations.GetOrganizationsResult {
+    return organizations.GET_ORGANIZATIONS_RESULT.is(result);
+  }
```
This file is inside a library, which should be usable across multiple projects inside this monorepo. I am very certain that async autocomplete field will find its use in RPM rewrite.

Therefore, this component should not know about various different return values and tailor its behavior accordingly. Make this component generic and if you need a specific behavior, make a wrapper around it.

Pl217 commented on libs/hpc-ui/src/lib/components/form-fields/async-autocomplete-field.tsx:
```diff
@@ -0,0 +1,247 @@
+  category,
+  isAutocompleteAPI,
+  ...otherProps
+}: AsyncAutocompleteSelectProps) => {
```
In an effort to make this component generic, which I wrote about in other comment inside this file, we can make this a generic and not specify API result types.
```ts
interface AsyncAutocompleteSelectProps<R> {
  ...
  fnPromise: ({ query }: { query: string }) => Promise<R[]>;
  ...
};

// Then use this interface, but notice how we specify generic with a trailing comma
const AsyncAutocompleteSelect = <R,>({
  ...
  fnPromise,
  ...
}: AsyncAutocompleteSelectProps<R>) => {
  ...
}

// Then, when you want to use this component
<C.AsyncAutocompleteSelect<locations.GetLocationsAutocompleteResult[number]>
  fnPromise={environment.model.locations.getAutocompleteLocations}
/>
```

unocha-hpc reviewed (approved):
Checks have passed and this pull request is ready for manual review

Pl217 reviewed (changes requested):
Here are some of the initial comments I have. I just reached 95332f8c454dd334332b1f7e765322be933ec803 and reviewed until file `async-autocomplete-field.tsx`. There was a problem when I ran `npm install`, so I found what the issue was and resolved it in a separate commit that I pushed.

Onitoxan committed:
Remove yup and add io-ts support

Onitoxan committed:
Update files from .tsx to .ts

Onitoxan committed:
Remove unnused imports and minor fixes

unocha-hpc reviewed (approved):
Checks have passed and this pull request is ready for manual review

unocha-hpc reviewed (approved):
Checks have passed and this pull request is ready for manual review

manelcecs committed:
Fix wrong type for next/prev page cursor

unocha-hpc reviewed (approved):
Checks have passed and this pull request is ready for manual review

manelcecs committed:
Update nested flow filters

Fix io-ts search form validations

unocha-hpc reviewed (approved):
Checks have passed and this pull request is ready for manual review

manelcecs committed:
Modify (i) icon mouseover from Source and Destination organizations

Add preliminary adjustment to avoid filtering by source and destination on collections from GraphQL

unocha-hpc reviewed (approved):
Checks have passed and this pull request is ready for manual review

manelcecs committed:
Add 'source system ID' column to flow table

Added translations for table headers. Also improved how default table headers where habdled the first time

manelcecs committed:
Add 'reporter reference code' column to flow table

Added translations for table headers.

unocha-hpc reviewed (approved):
Checks have passed and this pull request is ready for manual review

manelcecs committed:
Correct usage of column name

unocha-hpc reviewed (approved):
Checks have passed and this pull request is ready for manual review

unocha-hpc reviewed (approved):
Checks have passed and this pull request is ready for manual review

manelcecs committed:
Add download flows with fetch

New validation for download

Onitoxan committed:
Remove unused params on dummy.ts

Onitoxan committed:
Refactor AsyncAutocompleteSelect and add i18n support

unocha-hpc reviewed (approved):
Checks have passed and this pull request is ready for manual review

Pl217 commented on apps/hpc-ftsadmin/src/app/components/filters/filter-flows-REST-table.tsx:
```diff
@@ -0,0 +1,475 @@
+import tw from 'twin.macro';
+import React, { useState } from 'react';
+
+import * as io from 'io-ts';
```
We always import `io-ts` with `import * as t from 'io-ts'`, so using `io` instead of `t` has confused me.

Onitoxan replied:
Yeah I know, but through all frontend code we also import:
`import { t } from '../../i18n';`
So I decided to import it as `io` for not causing problems or further confusion

Pl217 commented on apps/hpc-ftsadmin/src/app/components/filters/filter-flows-REST-table.tsx:
```diff
@@ -0,0 +1,475 @@
+  };
+
+  const FORM_VALIDATION = io.partial({
+    flowID: io.array(codecs.INTEGER_FROM_STRING),
```
`INTEGER_FROM_STRING` will accept negative numbers too.

Pl217 commented on apps/hpc-ftsadmin/src/app/utils/form-validation.ts:
```diff
@@ -0,0 +1,33 @@
+    const errors: FormikErrors<T> = {};
+    result.left.map((value) =>
+      value.context.map((context) => {
+        if (isKey(values, context.key)) {
```
You are using `Array.prototype.map()` instead of a `for...of` loop. Mapping means iterating values and creating new array elements of it, which you aren't doing. Rewrite this using regular loops.

Pl217 commented on apps/hpc-ftsadmin/src/app/utils/form-validation.ts:
```diff
@@ -0,0 +1,33 @@
+        if (isKey(values, context.key)) {
+          const key = context.key;
+          // eslint-disable-next-line @typescript-eslint/ban-ts-comment
+          //@ts-ignore
```
Don't use `@ts-ignore`

Pl217 commented on apps/hpc-ftsadmin/src/app/utils/form-validation.ts:
```diff
@@ -0,0 +1,33 @@
+          const key = context.key;
+          // eslint-disable-next-line @typescript-eslint/ban-ts-comment
+          //@ts-ignore
+          errors[key] = '{validationError}'; //  Placeholder to change for i18n text
```
I've searched the codebase to find where you replace `validationError` placeholder and found out that you aren't doing it.

Pl217 commented on apps/hpc-ftsadmin/src/app/components/filters/filter-flows-REST-table.tsx:
```diff
@@ -0,0 +1,475 @@
+    setTableInfoDisplay(false);
+  };
+
+  const FORM_VALIDATION = io.partial({
```
In 6bd3f1501af14d3ff9fe4559715a5274330489a0, where you replace `yup` codecs with `io-ts` codecs, you have removed lots of properties from this codec like `keywords`, `flowStatus`, `sourceOrganizations`, `sourceLocations`, `sourceUsageYears`, `destinationLocations`, `destinationUsageYears`.

Pl217 commented on apps/hpc-ftsadmin/src/app/components/filters/filter-flows-table.tsx:
```diff
@@ -0,0 +1,505 @@
+  reporterRefCode: io.union([codecs.NON_EMPTY_STRING, codecs.EMPTY_STRING]),
+  sourceSystemID: io.union([codecs.NON_EMPTY_STRING, codecs.EMPTY_STRING]),
+  legacyID: io.union([codecs.INTEGER_FROM_STRING, codecs.EMPTY_STRING]),
+});
```
Before switching from `yup` to `io-ts` in 6bd3f1501af14d3ff9fe4559715a5274330489a0, this codec used to have lots of source/destination properties. Are you sure we don't need them?

Onitoxan replied:
I think they were not really necessary, while starting to do validations using yup, I tried to see how to validate these object values (In the code I call them FormObjectValue). When doing the refactor, I realised there is no way on the frontend to insert any other possible value, so I thought they were only necessary for backend validation.

To make a summary, I decided to remove the non useful validations.

Pl217 commented on apps/hpc-ftsadmin/src/app/components/filters/filter-organization-table.tsx:
```diff
@@ -0,0 +1,193 @@
+    ORGANIZATIONS_FILTER_INITIAL_VALUES
+  );
+
+  const FORM_VALIDATION = io.partial({
+    organization: io.string,
+  });
```
Before switching to `io-ts` in 6bd3f1501af14d3ff9fe4559715a5274330489a0, this codec had a lot more properties: `parentOrganization`, `locations`, `organizationType` and `status` are missing now.

Onitoxan replied:
I think they were not really necessary, while starting to do validations using yup, I tried to see how to validate these object values (In the code I call them FormObjectValue). When doing the refactor, I realised there is no way on the frontend to insert any other possible value, so I thought they were only necessary for backend validation.

To make a summary, I decided to remove the non useful validations.

Pl217 commented on libs/hpc-ui/src/lib/components/form-fields/async-autocomplete-field.tsx:
```diff
@@ -0,0 +1,168 @@
+        {...params}
+        size="small"
+        label={label}
+        required={requiered}
```
Isn't it obvious here that you've made a typo? Fix all the usages of word "requiered"

Pl217 commented on libs/hpc-data/src/lib/flows.ts:
```diff
@@ -0,0 +1,503 @@
+    sourceSystemID: t.string,
+  }),
+});
+// TODO: transform to arrays of
```
Arrays of what? This TODO doesn't have a meaning.

Onitoxan replied:
No idea, removed

Pl217 commented on libs/hpc-live/src/lib/utils.ts:
There a bunch of `if` statements in this file, that aren't using curly braces in their then branches. This would fail ESLint's [`curly`](https://eslint.org/docs/latest/rules/curly) linting rule, but we aren't using common rules from `hpc-repo-tools` in this repo yet, because of specifics of Nx, I haven't yet integrated it.

These aren't things I should be looking after in reviews, because that is job of a linter. So, you can keep current code and auto-fixing will change it once we integrate linting rules.

Pl217 commented on libs/hpc-data/src/lib/flows.ts:
```diff
@@ -0,0 +1,503 @@
+    childIDs: t.union([t.array(t.number), t.null]),
+    parentIDs: t.union([t.array(t.number), t.null]),
+    categories: t.union([t.array(FLOW_CATEGORY), t.null]),
+    // TO REMOVE
```
To remove when?

Onitoxan replied:
It has to do with removing the REST API of flows, if I don't remember badly I let comments like this around the code to remember where to look when removing things, at least for the search. In the main comment you commented if we could eliminate everything related to the REST endpoint, but Nick told me to maintain it there just to show the client.

In any case, we will keep the REST endpoint support and once this PR gets merged to develop, and the job of Add Flow merges to develop, we will remove the REST endpoint part, except the one needed for the Add Flows functionality.

Pl217 commented on apps/hpc-ftsadmin/src/app/components/tables/flows-table.tsx:
```diff
@@ -0,0 +1,900 @@
+                  if (row.reportDetails) {
+                    const uniqueSourceIDs = new Set(
+                      row.reportDetails.map((rd) => rd.sourceID)
+                    );
```
This produces `Set<string | null | undefined>` and values `null` and `undefined` won't be skipped when you convert `Set` to array and call `join(', ')` below. If you run `Array.from(['115114', undefined, '-68117;-67234', null]).join(', ')`, you will see that result is `'115114, , -68117;-67234, '`.

Therefore, let's filter out values not defined.
```suggestion
                    const uniqueSourceIDs = new Set(
                      row.reportDetails.map((rd) => rd.sourceID).filter(util.isDefined)
                    );
```

Pl217 commented on apps/hpc-ftsadmin/src/app/components/tables/flows-table.tsx:
```diff
@@ -0,0 +1,900 @@
+                    const uniqueRefCodesArray = Array.from(uniqueSourceIDs);
+                    rd = uniqueRefCodesArray.join(', ');
+                    rd = rd.length > 0 ? rd : '--';
+                  }
```
You have almost the same construct as above and same problems I wrote about there, regarding `undefined` and `null`, also apply here.

Pl217 commented on libs/hpc-data/src/lib/flows.ts:
```diff
@@ -0,0 +1,503 @@
+  typeof GET_TOTAL_AMOUNT_USD_PARAMS
+>;
+
+export const GET_TOTAL_AMOUNT_USD_RESULT = t.type({
+  searchFlowsTotalAmountUSD: t.type({
+    totalAmountUSD: t.string,
+  }),
+});
+export type GetTotalAmountUSDResult = t.TypeOf<
+  typeof GET_TOTAL_AMOUNT_USD_RESULT
+>;
```
Looks like lots of nesting just to represent a single string amount.

Onitoxan replied:
Well, TotalAmountUSD is giving problems because it consumes a lot, and since is not feature parity, we will take it out for the moment. So this will answer this comment

Pl217 commented on apps/hpc-ftsadmin/src/app/utils/fn-promises.ts:
```diff
@@ -0,0 +1,91 @@
+    };
+  });
+  return options;
+};
```
You can return the mapped array directly, without storing it to a variable first. Also, you can simplify returns in arrow functions. For slightly less verbose code, you cal apply similar pattern across this file:
```suggestion
export const fnOrganizations = async (
  query: { query: string },
  env: Environment
) => {
  const response =
    await env.model.organizations.getAutocompleteOrganizations(query);

  return response.map((responseValue) => ({
    displayLabel: `${responseValue.name} [${responseValue.abbreviation}]`,
    value: responseValue.id,
  }));
};
```

Pl217 commented on libs/hpc-ui/src/lib/components/form-fields/async-multi-select-field.tsx:
```diff
@@ -0,0 +1,156 @@
+        label={label}
+      />
+    ),
+    renderValue: (selected: any) => (
```
Don't use `any`, it just masks the typing problems.

Onitoxan replied:
When removing the `any` there was no typing problem (I refactored all UI components and probably forgot to remove the any). Even worse than masking a problem, it was creating an non existent one. Removed

Pl217 commented on libs/hpc-ui/src/lib/components/form-fields/async-single-select-field.tsx:
```diff
@@ -0,0 +1,157 @@
+  preOptions?: Array<FormObjectValue>;
+};
+
+/** FIXME: Component miss-beheaving */
```
What is the misbehavior? Also, this is a typo, word is "misbehaving" :)

Onitoxan replied:
I was trying to use a component that is not supposed to work as I expected, so I removed it, all has been replaced for `async-autocomplete` that allow in fewer lines of code to achieve the result I wanted.

Pl217 commented on libs/hpc-ui/src/lib/components/form-fields/async-single-select-field.tsx:
```diff
@@ -0,0 +1,157 @@
+          options.map((value) => (
+            //  TODO: Find a way to add { label:string, value:string } as value, maybe by enconding it to a string or any other method
+            // eslint-disable-next-line @typescript-eslint/ban-ts-comment
+            //@ts-ignore
```
Don't use `@ts-ignore`

Pl217 commented on libs/hpc-ui/src/lib/components/form-fields/async-single-select-field.tsx:
There are a few TODO items in this file which we should fix before merging this.

Pl217 commented on libs/hpc-ui/src/lib/components/form-fields/autocomplete-field.tsx:
```diff
@@ -0,0 +1,73 @@
+  isMulti?: boolean;
+}) => {
+  const { setFieldValue } = useFormikContext();
+  const [field, meta] = useField(name);
```
`meta` is unused
```suggestion
  const [field] = useField(name);
```

Pl217 commented on libs/hpc-ui/src/lib/components/form-fields/autocomplete-field.tsx:
```diff
@@ -0,0 +1,73 @@
+    options: selectOptions,
+    getOptionLabel: (op) => op,
+    ChipProps: { size: 'small' },
+    onInputChange: (event, newInputValue) => handleInputChange(newInputValue),
```
`event` is unused
```suggestion
    onInputChange: (_, newInputValue) => handleInputChange(newInputValue),
```

Pl217 commented on libs/hpc-ui/src/lib/components/form-fields/checkbox.tsx:
```diff
@@ -0,0 +1,52 @@
+  name: string;
+  label?: string;
+  value?: any;
+  onChange?: (event: React.ChangeEvent<HTMLInputElement>) => void | any;
```
Don't use `any`. The argument `value`doesn't seem to be used for its value (only checking existance), so you can just use `unknown`. As for `onChange` maybe you can use `string` as return value, but I am not sure.

Pl217 commented on libs/hpc-ui/src/lib/components/form-fields/date-picker.tsx:
```diff
@@ -0,0 +1,97 @@
+  lang?: string;
+  enableButton?: boolean;
+}) => {
+  const [field, meta, helpers] = useField(name);
```
`meta` is unused.

Pl217 commented on libs/hpc-ui/src/lib/components/form-fields/single-select-field.tsx:
```diff
@@ -0,0 +1,98 @@
+          <MenuItem
+            key={`${value.displayLabel}_${value.value}`}
+            value={value as any}
+          >
```
Type casting like this is dangerous. Don't use it

Pl217 commented on libs/hpc-ui/src/lib/components/form-fields/text-field.tsx:
```diff
@@ -0,0 +1,86 @@
+    InputProps:
+      type === 'number' || type === 'currency'
+        ? {
+            inputComponent: NumericFormatCustom as any,
```
What are you hiding with this type cast? :P

Onitoxan replied:
I was hiding a monster, but it has been solved eheh

Pl217 commented on libs/hpc-ui/src/lib/components/draggable-list.tsx:
Remove usages of `any` in this file

Pl217 commented on apps/hpc-ftsadmin/src/app/paths.ts:
```diff
@@ -1,5 +1,13 @@
 const HOME = '/';
+const ROOT = '/*';
 const FLOWS = '/flows';
+const FLOWS_REST = '/flowsREST'; // TODO: Remove once shown to client
```
Are we ready to remove 2 TODO items in this file, or we need to keep them for client feedback?

Pl217 commented on apps/hpc-ftsadmin/src/app/components/tables/flows-table-REST.tsx:
```diff
@@ -0,0 +1,775 @@
+    if (!firstRender) {
+      load();
+    } else {
+      firstRender = false;
```
There is a warning on this line: Assignments to the 'firstRender' variable from inside React Hook useEffect will be lost after each render

Onitoxan replied:
All things related to the file flows-table-REST.tsx will not be done.

This file should be deleted, I was waiting for Nick's approval, but I think In further commits fixing this PR I will get rid of it. We have it for the moment in red.dev, and of course, in the commit history if needed. I will stop maintaining it.

Pl217 commented on apps/hpc-ftsadmin/src/app/components/tables/flows-table-REST.tsx:
```diff
@@ -0,0 +1,775 @@
+    }
+  };
+
+  const handleChangePage = (event: unknown, newPage: number) => {
```
`event` is unused

Onitoxan replied:
All things related to the file flows-table-REST.tsx will not be done.

This file should be deleted, I was waiting for Nick's approval, but I think In further commits fixing this PR I will get rid of it. We have it for the moment in red.dev, and of course, in the commit history if needed. I will stop maintaining it.

Pl217 commented on apps/hpc-ftsadmin/src/app/components/tables/flows-table-REST.tsx:
```diff
@@ -0,0 +1,775 @@
+              </TableCell>
+            )}
+            {decodeTableHeaders(query.tableHeaders, lang).map((column) => {
+              if (!column.active) {
+                return;
+              }
```
There is a warning "Array.prototype.map() expects a return value from arrow function", which you can resolve by filtering inactive values before mapping over them.

You can use `filter((header) => header.active)` here and in one more place below.

Onitoxan replied:
All things related to the file flows-table-REST.tsx will not be done.

This file should be deleted, I was waiting for Nick's approval, but I think In further commits fixing this PR I will get rid of it. We have it for the moment in red.dev, and of course, in the commit history if needed. I will stop maintaining it.

Pl217 commented on apps/hpc-ftsadmin/src/app/components/tables/flows-table-REST.tsx:
```diff
@@ -0,0 +1,775 @@
+                          setQuery({
+                            ...query,
+                            tableHeaders: encodeTableHeaders(
+                              element as any, // TO DO: remove any
```
Do the TODO

Onitoxan replied:
All things related to the file flows-table-REST.tsx will not be done.

This file should be deleted, I was waiting for Nick's approval, but I think In further commits fixing this PR I will get rid of it. We have it for the moment in red.dev, and of course, in the commit history if needed. I will stop maintaining it.

Pl217 commented on apps/hpc-ftsadmin/src/app/components/tables/flows-table.tsx:
```diff
@@ -0,0 +1,900 @@
+    newPage: number,
+    prevPageCursor: number,
+    nextPageCursor: number
+  ) => {
```
`event` is unused

Pl217 commented on apps/hpc-ftsadmin/src/app/components/tables/flows-table.tsx:
```diff
@@ -0,0 +1,900 @@
+                            setQuery({
+                              ...query,
+                              tableHeaders: encodeTableHeaders(
+                                element as any, // TO DO: remove any
```
Do the TODO

Pl217 commented on apps/hpc-ftsadmin/src/app/components/tables/keywords-table.tsx:
Resolve TODOs in this file.

Onitoxan replied:
All of them are solved now!, while working on it, I had to diverse to work on proper component to display errors, fixing components, refactoring and many more.

Pl217 commented on apps/hpc-ftsadmin/src/app/utils/table-headers.ts:
```diff
@@ -0,0 +1,443 @@
+}
+export const DEFAULT_ORGANIZATION_TABLE_HEADERS: TableHeadersProps<OrganizationHeaderID>[] =
+  [];
+for (const [index, header] of Object.entries(POSSIBLE_ORGANIZATION_VALUES)) {
```
```suggestion
for (const [, header] of Object.entries(POSSIBLE_ORGANIZATION_VALUES)) {
```

Pl217 reviewed (changes requested):
- When switching from `yup` to `io-ts`, you have left out some properties of `yup` codecs and removed them from `io-ts`. I've commented about this in few places, but not all. [X] (Commented on one of the comments the reason)
- When removing `dependencies` in 5d6a12a58b6a1f44583b25e0c0487579806e6c08, you haven't updated lockfile. [X]
- Please update initial comment of this PR and write a proper PR message
- The number of warnings about missing dependencies in `useEffect` hooks has increased
- Do we need to keep both GraphQL and REST versions? [X] Removed
- Remove usages of type casting (`as` keyword). I did write about this in few places inline, but not everywhere [X] (I let in some places because of necessity)

unocha-hpc reviewed (approved):
Checks have passed and this pull request is ready for manual review

Onitoxan committed:
Add io-ts codec to accept possitive values

unocha-hpc reviewed (approved):
Checks have passed and this pull request is ready for manual review

Onitoxan committed:
Fix typo 'requiered'

Onitoxan committed:
Remove strange TODO comment

unocha-hpc reviewed (approved):
Checks have passed and this pull request is ready for manual review

Onitoxan committed:
Add filtering of non-defined values on Set

unocha-hpc reviewed (approved):
Checks have passed and this pull request is ready for manual review

Onitoxan committed:
Update fn-promises to be less verbose

unocha-hpc reviewed (approved):
Checks have passed and this pull request is ready for manual review

Onitoxan committed:
Remove unnecessary 'any'

unocha-hpc reviewed (approved):
Checks have passed and this pull request is ready for manual review

unocha-hpc reviewed (approved):
Checks have passed and this pull request is ready for manual review

Onitoxan committed:
Remove unused values from autocomplete-field

unocha-hpc reviewed (approved):
Checks have passed and this pull request is ready for manual review

unocha-hpc reviewed (approved):
Checks have passed and this pull request is ready for manual review

Onitoxan committed:
Remove unnecessary 'any' and dangerous type assertions

Onitoxan committed:
Remove unused values

unocha-hpc reviewed (approved):
Checks have passed and this pull request is ready for manual review

Onitoxan committed:
Remove REST flow page

unocha-hpc reviewed (approved):
Checks have passed and this pull request is ready for manual review

Onitoxan committed:
Fix form-validation

unocha-hpc reviewed (approved):
Checks have passed and this pull request is ready for manual review

Onitoxan committed:
Remove TotalAmountUSD

Onitoxan committed:
Remove unnecessary components and update async-autocomplete

Onitoxan committed:
Remove any from draggable-list

Onitoxan committed:
Solve TODOS of keyword-table and related issues

Onitoxan committed:
Refactors, Fixes and minor updates

unocha-hpc reviewed (approved):
Checks have passed and this pull request is ready for manual review

unocha-hpc reviewed (approved):
Checks have passed and this pull request is ready for manual review

unocha-hpc reviewed (approved):
Checks have passed and this pull request is ready for manual review

Onitoxan committed:
Fix organization filter validation

unocha-hpc reviewed (approved):
Checks have passed and this pull request is ready for manual review

unocha-hpc reviewed (approved):
Checks have passed and this pull request is ready for manual review

Pl217 commented on package.json:
We install a ton of new libraries in the first commit, likely increasing the bundle size by a lot, but I didn't check this.

Let's use the latest version for all newly added packages. Also, when specifying versions, use exact version, without caret (^ symbol), because that is what the majority of dependencies do.

Pl217 commented on package.json:
```diff
@@ -44,9 +44,10 @@
     "@mui/material": "^5.12.2",
     "@mui/styled-engine-sc": "^5.12.0",
-    "@nx/cypress": "18.3.0",
+    "@mui/x-date-pickers": "^6.16.2",
```
I see that you have removed `@nx/cypress`. Why is that? You seem to have done this by mistake, since the change is not reflected in lockfile.

Also, there is a next major version of `@mui/x-date-pickers`, so please upgrade.

Pl217 commented on package.json:
```diff
@@ -87,6 +93,8 @@
     "eslint-plugin-react": "7.32.2",
     "eslint-plugin-react-hooks": "4.6.0",
     "express": "4.19.2",
+    "file-saver": "^2.0.5",
```
I see that `file-saver` isn't regularly maintained and hasn't seen a release since August 2020, so we can likely remove it, since it seems unused.

Pl217 commented on libs/hpc-ui/postcss.config.js:
```diff
@@ -0,0 +1,16 @@
+    tailwindcss: {
+      config: join(__dirname, 'tailwind.config.js'),
+    },
+    autoprefixer: {},
```
What does this do? Did we install `autoprefixer` and not using it?

Onitoxan replied:
To use tailwind as PostCSS plugin I need to add autoprefixer to the config file:

https://tailwindcss.com/docs/installation/using-postcss

unocha-hpc reviewed (approved):
Checks have passed and this pull request is ready for manual review

Pl217 commented on libs/hpc-live/src/lib/utils.ts:
```diff
@@ -0,0 +1,112 @@
+function parseFilterToGraphQL(obj: Record<string, any>): string {
+  const keyValuePairs: string[] = [];
+
+  if (Array.isArray(obj)) {
```
Let's try and get rid of explicit `any` usage in this method. If I just replace argument definition with `Record<string, unknown>`, it immediately reveals problem obvious "with a naked eye" - and that is the fact that you're using `Record`, which is a utility type to define shape of an object, and in the first meaningful line of code, you check if it is an array.

Docstring above also states that this method can work with arrays, but you're only specifying an object in argument type definition.

Pl217 commented on libs/hpc-core/src/lib/i18n/index.ts:
```diff
@@ -106,11 +106,7 @@ export class LanguageChoice<LanguageKey extends string> {
-    } else {
-      document.documentElement.setAttribute('dir', meta.direction);
-    }
+    document.documentElement.setAttribute('dir', meta.direction);
```
What motivated this change?

Onitoxan replied:
When switching from LTR to RTL, and viceversa, if I didn't explicitly define `dir="ltr"` some MUI components would not apply properly styles on LTR, I think also with some css classes that use 'start' and 'end' also run into little problems (I changed this quite a long time ago, for MUI I'm sure, for css properties I don't remember if it was the case 100%)

Pl217 commented on libs/hpc-ui/src/lib/components/form-fields/async-autocomplete-field.tsx:
```diff
@@ -0,0 +1,165 @@
+
+const StyledAutocomplete = tw(Autocomplete)`
+    min-w-[10rem]
+    w-full`;
```
Not something we can enforce, but these definitely look nicer if they are formatted like rest of the code:
```suggestion
const StyledAutocomplete = tw(Autocomplete)`
  min-w-[10rem]
  w-full
`;
```

Pl217 commented on libs/hpc-ui/src/lib/components/form-fields/async-autocomplete-field.tsx:
```diff
@@ -0,0 +1,165 @@
+    return () => {
+      active = false;
+    };
+  }, [loading, inputValue]);
```
Linter complains and these warnings can also be found in GitHub Actions runs:
> React Hook useEffect has missing dependencies: 'data', 'fnPromise', and 'isAutocompleteAPI'. Either include them or remove the dependency array. If 'fnPromise' changes too often, find the parent component that defines it and wrap that definition in useCallback.

Pl217 commented on libs/hpc-ui/src/lib/components/form-fields/async-autocomplete-field.tsx:
```diff
@@ -0,0 +1,165 @@
+      setData([]);
+      setIsFetch(false);
+    }
+  }, [open]);
```
Linter complains and these warnings can also be found in GitHub Actions runs:
> React Hook useEffect has a missing dependency: 'isAutocompleteAPI'. Either include it or remove the dependency array.

Pl217 commented on libs/hpc-ui/src/lib/components/form-fields/async-autocomplete-field.tsx:
```diff
@@ -0,0 +1,165 @@
+    ...field,
+    ...otherProps,
+    multiple: isMulti,
+    placeholder: placeholder,
```
We can use more modern syntax (here and in a few places below):
```suggestion
    placeholder,
```

Pl217 commented on libs/hpc-ui/src/lib/components/form-fields/async-autocomplete-field.tsx:
```diff
@@ -0,0 +1,165 @@
+    },
+    onInputChange: (event, newInputValue) => {
+      setInputValue(newInputValue);
+    },
```
`event` is unused
```suggestion
    onInputChange: (_, newInputValue) => {
      setInputValue(newInputValue);
    },
```

Pl217 commented on libs/hpc-ui/src/lib/components/form-fields/autocomplete-field.tsx:
```diff
@@ -0,0 +1,65 @@
+
+const StyledAutocomplete = tw(Autocomplete)`
+      min-w-[10rem]
+      w-full`;
```
Would sure look nicer if these are formatted like rest of the code:
```suggestion
const StyledAutocomplete = tw(Autocomplete)`
  min-w-[10rem]
  w-full
`;
```

Pl217 commented on libs/hpc-ui/src/lib/components/form-fields/autocomplete-field.tsx:
```diff
@@ -0,0 +1,65 @@
+    ...otherProps,
+    multiple: isMulti,
+    readOnly,
+    options: options,
```
We can use more modern syntax
```suggestion
    options,
```

Pl217 commented on libs/hpc-ui/src/lib/components/form-fields/date-picker.tsx:
```diff
@@ -0,0 +1,97 @@
+
+const StyledLink = tw(Link)`
+ms-8
+`;
```
Would definitely be nicer if you applied consistent formatting in these
```suggestion
const StyledDatePicker = tw.div`
  flex
  w-full
  items-baseline
`;

const StyledLink = tw(Link)`
  ms-8
`;
```

Pl217 commented on libs/hpc-ui/src/lib/components/divider.tsx:
```diff
@@ -0,0 +1,11 @@
+border-b-unocha-pallete-gray-light  
+border-b
+my-8
+`;
```
```suggestion
const Divider = tw.div`
  border-solid
  border-0 
  border-b-unocha-pallete-gray-light  
  border-b
  my-8
`;
```

Pl217 commented on libs/hpc-ui/src/lib/components/section.tsx:
```diff
@@ -0,0 +1,68 @@
+      ? tw`p-4 hover:bg-unocha-panel-bgHover open:hover:bg-white`
+      : tw`p-0 w-full`,
+  ]
+);
```
This will be more readable
```suggestion
const SectionTitle = styled.summary(
  ({ type }: { type: 'primary' | 'secondary' }) => [
    tw`
      font-bold
      m-0
      transition-all
      ease-out
      duration-500
      list-none
      cursor-pointer
      [&::-webkit-details-marker]:hidden
    `,
    type === 'primary'
      ? tw`
        flex
        justify-between 
        uppercase
        font-bold
        text-2xl
        text-unocha-pallete-blue-dark2
      `
      : tw`text-right [&>svg]:ms-2 text-unocha-pallete-blue text-xl`,
  ]
);
```

Pl217 commented on libs/hpc-ui/src/lib/styling.tsx:
```diff
@@ -49,6 +49,15 @@ body {
+  width:1px;
+  height:1px;
+  overflow:hidden;
+}
```
I've made lots of comments on formatting on these and that isn't something a review should be concerned with. The big downside with the idea of Tailwind (with CSS-in-JS) is that we don't get formatting with the tooling we're using.

In this file, all the CSS statements are having a space after colon, but you didn't use it
```suggestion
.${CLASSES.VISUALLY_HIDDEN} {
  position: absolute;
  left: -10000px;
  top: auto;
  width: 1px;
  height: 1px;
  overflow: hidden;
}
```

Pl217 commented on apps/hpc-ftsadmin/src/app/utils/download-excel.ts:
```diff
@@ -0,0 +1,18 @@
+//  TODO: Check if this should be done on backend side, Currently not working propperly
```
You've got a typo

Pl217 commented on apps/hpc-ftsadmin/src/app/utils/map-functions.ts:
Many methods in this file are not used.

Pl217 commented on package.json:
```diff
@@ -96,18 +104,31 @@
     "ts-node": "10.9.1",
-    "typescript": "5.2.2"
+    "typescript": "5.2.2",
+    "twin.macro": "^3.3.1",
```
`typescript` comes chronologically after `twin.macro`. You must have changed this manually.

Pl217 commented on apps/hpc-ftsadmin/src/app/utils/table-headers.ts:
```diff
@@ -0,0 +1,501 @@
+
+/** The nomenclature to define these IDs is to write it like: name of the DB table, and after the dot, the propperty.
+ *  if it's not a DB field, just write down the name
+ */
```
Not sure why Prettier doesn't enforce length of lines in docstrings and proper spacing (to avoid double spaces), but this looks so messy. Also, you've got the typo in there.
```suggestion
/**
 * The nomenclature to define these IDs is to write it like:
 * name of the DB table, and after the dot, the property.
 * If it's not a DB field, just write down the name.
 */
```

Pl217 commented on apps/hpc-ftsadmin/src/app/pages/flows/flows-list.tsx:
```diff
@@ -0,0 +1,130 @@
+    return () => {
+      handleAbortController();
+    };
+  }, []);
```
Linter complains and these warnings can also be found in GitHub Actions runs:
> React Hook useEffect has a missing dependency: 'handleAbortController'. Either include it or remove the dependency array

Pl217 replied:
So, you decided to ignore it 🫣

Onitoxan replied:
(Same as comment below)

Pl217 commented on apps/hpc-ftsadmin/src/app/pages/flows/pending-flows-list.tsx:
```diff
@@ -0,0 +1,129 @@
+    return () => {
+      handleAbortController();
+    };
+  }, []);
```
Linter complains and these warnings can also be found in GitHub Actions runs:
> React Hook useEffect has a missing dependency: 'handleAbortController'. Either include it or remove the dependency array

Pl217 replied:
So, you decided to ignore it 🫣

Onitoxan replied:
Okay so for this, it may be a lack of knowledge from my part in react, but basically, I want to just execute when the component gets unmounted.  My inner guts tells me this is not the way to go, because `useEffect` in general shouldn't be used for this cases but its in my head of old ways to use react when people used classes to be able to define what to do `onUnmount`.

If I write handleAbortController it will update if handleAbortController changes (what it shouldn't) but in any case I just want this to be updated onUnmount, and it could be missleading to say that is depending on change of `handleAbortController`

What do you think is the real proper solution, because this warning even if I solve it in other places, it just gives me problems.

Pl217 replied:
Honestly, linter warnings around the dependencies of `useEffect` are often giving me problems too. I don't always know how to resolve it, but I am not too skilled with frontend development and React.

In this case, I extracted the code pieces from this component and asked Claude AI models for help. After some back & forth, it offered me what sounds a sensible solution, but take it with a grain of salt, because AI models hallucinate too often.

The main point is that your `handleAbortController` is aborting the request and then creating a new `AbortController`. But, when component is unmounting, you just need to abort the request, there is no point in creating a new `AbortController`. So, if we just call `abortController.abort()` in the `useEffect` handler, instead of entire `handleAbortController`, that should be enough.

AI bot also suggested switching from `useState` to `useRef`, presumably because we would not need to add it to dependencies of `useEffect`, but I don't know, because it doesn't warn me on local and I cannot even find annotations in the GitHub Actions runs that I wrote about originally. Something may be broken with linting configuration.

Anyway, this is the code it produced.
```tsx
const abortControllerRef = useRef<AbortController>(new AbortController());

// const [query, setQuery] = useQueryParams({ ... });

const handleAbortController = useCallback(() => {
  abortControllerRef.current.abort();
  abortControllerRef.current = new AbortController();
}, []);

useEffect(() => {
  return () => {
    abortControllerRef.current.abort();
  };
}, []);

const pendingFlowsTableProps: FlowsTableProps = {
  headers: DEFAULT_FLOW_TABLE_HEADERS,
  initialValues: PENDING_FLOWS_FILTER_INITIAL_VALUES,
  rowsPerPageOption: [10, 25, 50, 100],
  query: query,
  setQuery: setQuery,
  pending: true,
  abortSignal: abortControllerRef.current.signal,
};
```

Pl217 reviewed (changes requested):
Good job! I really like how you've learned Typescript (generic, type guards, etc.) and some other more advanced concepts I've seen in the code. For the review, I have transformed myself into linter/formatter and complained a little ;)

---

- [x] I did make too many inline comments regarding the formatting of Tailwind classes with CSS-in-JS, but stopped at some point. Apply the same formatting style to all places where you find results for search terms "tw(" and "tw.".

This isn't a job for humans, but I didn't succeed in configuring prettier to do this for us. Prettier natively supports CSS-in-JS [since version 1.5.0](https://prettier.io/blog/2017/06/28/1.5.0) and it works with styled-components. But, this applies to CSS rules, which have key-value syntax, with a colon. Each rule would be on its own line, ending with a semicolon. But, you now added Tailwind with styled-component-like syntax, where you are specifying class names, not CSS rules. And Prettier cannot handle this. I have tried installing [prettier-plugin-tailwindcss](https://www.npmjs.com/package/prettier-plugin-tailwindcss) and adding this to `.prettierrc`:
```json
"plugins": ["prettier-plugin-tailwindcss"],
"tailwindFunctions": ["tw"]
```
but it only tries to apply sorting of classes in some cases, doesn't do the formatting. Try finding a plugin/tool that will check the formatting for us. See if `twin.macro` suggests something on this, but I didn't find anything on a quick search.

---

- [x] Also, I have made a few comments on how you can use more modern JS syntax, where you don't have to type `property: property` if names match. There are definitely more places where this same suggestion applies, but I won't comment on all of them.

---

- [x] Also, the way you write docstrings is weird. I don't know why Prettier doesn't standardize things like newlines, spaces and line length, but I left one inline comment to indicate how they are usually written.

---

- [x] The commit 419654ff0511b89520f419c22b17cf05eed811d3 says "Remove 'dependencies' from package.json", but it never touches `package.json`, only lockfile.

---

- [x] Update the title of the PR not to only mention ticket number.

Onitoxan committed:
Fix CSS-in-JS formating for twin.macro

Onitoxan committed:
Refactor searchFlowParams() to be shorter and more comprenhensible

Onitoxan committed:
Fix typo and formatting issues

Onitoxan committed:
Remove unused functions

Onitoxan committed:
Prettify styling.tsx

Onitoxan committed:
Fix linting warnings

Onitoxan committed:
Disable eslint in required files

I just want this code to run on unmount and not everytime handleAbortController() re-renders

Onitoxan committed:
Refactor using modern ES6 features

unocha-hpc reviewed (approved):
Checks have passed and this pull request is ready for manual review

unocha-hpc reviewed (approved):
Checks have passed and this pull request is ready for manual review

Pl217 commented on libs/hpc-ui/src/lib/components/form-fields/autocomplete-field.tsx:
```diff
@@ -0,0 +1,66 @@
+  placeholder?: string;
+  readOnly?: boolean;
+  isMulti?: boolean;
+}) => {
```
I thought `...otherProps` would allow any additional properties which aren't directly defined. So, I tried adding a new field to the first place I found this component is used - in `organization-form.tsx`. It looked like this
```tsx
<C.AutocompleteSelect
  ...
  name="organizationLevel"
  // This is newly added property
  unspecifiedProperty="itsValue"
  ...
/>
```
I added `unspecifiedProperty`, but I get a type error.

I see there is a similar pattern of using `...otherProps` throughout the codebase, in other components too. Let's reconsider why you added it and remove if it does nothing.

As for this particular component (`AutocompleteSelect`), the usages are specifying only required props - `name`, `label`, `options` and one usage where you specify `readOnly` too. Reconsider whether you need `placeholder` and `isMulti` if they won't be used.

If you decide to tidy, do so for other components as well, not only `AutocompleteSelect` where I am writing this comment.

Onitoxan replied:
For this comment, in the case of `AutocompleteSelect` There are some props that are not used in this branch, and in the rest of my work I didn't use it either, so I removed them from the component 👍 , I thought that even if I didn't use them it couldn't hurt to let it implemented since it was easy to do, but I understand that when in the future we need (for example) `multiple` AutocompleteSelect we should build it for it to satisfy the problem we face at the moment.

For the rest of inputs, there some unused props, but in my work on top of this branch I end up using them, so i didn't remove them.

For this kind of PR I think we should be a bit more flexible since is not finished work, once I finish I will review myself everything and in that moment I think is nice if you are very picky so everything can be addressed because it has to be.

Pl217 commented on apps/hpc-ftsadmin/src/app/components/tables/keywords-table.tsx:
```diff
@@ -0,0 +1,594 @@
+    const traducedError = t.t(
+      lang,
+      (s) => s.components.keywordTable.errors[errorCode]
+    );
```
I was wondering what "traduced" error means. That is an unknown word to me. More importantly, you have pretty similar method `parseError` in `organization-form.tsx`. They differ in `error` argument. Could the functionality be extracted into a common place where both components can use it?

Pl217 commented on package.json:
```diff
@@ -87,27 +93,41 @@
+    "twin.macro": "3.4.1",
+    "typescript": "5.2.2",
+    "use-query-params": "2.2.1",
+    "xlsx": "0.18.5",
```
Recently, I found out that this library is no longer published via NPM, for the weirdest reasons possible. Look at ticket HPC-9693, where I explain the details and suggest removing it from Project Module. So, let's not add it here.

The only place where this lib is used (`download-excel.ts`) even says it's not working properly. So, either fix it using another library, or just remove the feature (and open ticket to implement in separate PR) and the library.

Onitoxan replied:
I replaced it, the thing is that in following commits of other branches I actually implemented the logic for this file and functionality. Anyways Ill replace this basic case in this ticket and when I put time to rebase HPC-9502 I will re-implement the features with the new library

Pl217 commented on libs/hpc-data/src/lib/flows.ts:
```diff
@@ -0,0 +1,337 @@
+  name: t.string,
+});
+
+/** Delete when finishing off REST flow endpoint (Maybe Matyas needs it) */
```
Do we still need to keep around REST code?

Onitoxan replied:
(Since this code is old, in the end we need it for add/edit flow, in my latest commits in other branches I reviewed this codec because it was not well defined either)

So we let it like this, in other tickets it's addressed

Pl217 replied:
Can we at least not mention former developers? I'd just delete entire comment

Pl217 commented on package.json:
```diff
@@ -87,27 +93,41 @@
+    "typescript": "5.2.2",
+    "use-query-params": "2.2.1",
+    "xlsx": "0.18.5",
+    "yup": "1.4.0"
```
You forgot to remove Yup.

Pl217 committed:
Format comments

Pl217 reviewed (changes requested):
Let's get this thing off the ground. It's been a long time since this was opened and I went through the history of our conversation and the changes asked in my previous requests. I see that this PR is now in a much better state. If I were to dig deeper, I'd likely find something more, but I think that you too would change many things in your own code after this much time.

So, I've got some minor things inline that we can change quickly, then we can merge this and make further work easier, since you don't need to perform many rebases on top of unmerged branches anymore. Please try and clean up any unused code (like REST code).

unocha-hpc reviewed (approved):
Checks have passed and this pull request is ready for manual review

unocha-hpc reviewed (approved):
Checks have passed and this pull request is ready for manual review

Pl217 commented:
I have reorganized commits a little bit and made the following changes:
- Squashed all commits that change `package.json` into the first commit. I have installed packages properly there, since you used to remove lockfile before reinstalling packages, which messed with `package-lock.json`
- Formatted all Tailwind usages properly, because you messed up in few places. Again, this isn't job for humans, but I noticed during review and decided to set a good example, but I cannot keep playing the formatter in the future
- Formatted code comments in few places I noticed. As I wrote in my reviews, your ways of writing comments looks weird in places, especially with docstrings, which are meant to comment on a file or code block, but docstring actually documents what next identifier (like method, variable, type, etc) does. I have formatted some of the comments, but when we introduce common linting rules to this repo, like we have in other repos, we should be able to mass-fix using ESLint

I have rebased on top of latest `develop` and then rebased branches `HPC-9502` (PR #439) and `HPC-9578` (PR #462) on top. Make sure you pull the latest code and not work on outdated local version.

There is a problem in hpc_service introduced by this PR. You told me that it is fixed in next commit (#439), so make sure to backport the fix, since we don't know when next PR will get merged and we cannot afford to break local API because of changes from this PR.

**Edit:** Actually, the problem isn't fixed in #439 either. The problem started appearing because in commit a6f89c673bfee16bf3975fa2bd7aaf77ebf6249c, you renamed io-ts codec `INTEGER_FROM_STRING` to `POSITIVE_INTEGER_FROM_STRING`, but it is used elsewhere with its former name. The codecs defined in CDM are used throughout hpc_service (and that is main reason CDM is checked out in hpc_service), and you can see this effect if you check out this branch and run `npm run check-types` in hpc_service.

unocha-hpc reviewed (approved):
Checks have passed and this pull request is ready for manual review

unocha-hpc reviewed (approved):
Checks have passed and this pull request is ready for manual review

Onitoxan committed:
Remove unnecessary '...otherProps'

unocha-hpc reviewed (approved):
Checks have passed and this pull request is ready for manual review

Pl217 reviewed (changes requested):
Since my last review, you have removed all usages of `...otherProps` in 84071987a05b368d5ec576895c0055887d584756 and resolved the problem where you overwritten codec `INTEGER_FROM_STRING` with `POSITIVE_INTEGER_FROM_STRING`.

But, you ignored everything else I wrote without commenting on it. I asked to remove any unused code, like REST code and like unused component properties, which I wrote about in [this comment](https://github.com/UN-OCHA/hpc-cdm/pull/420#discussion_r1740448245). Looking back now at that comment, I see that file is named `autocomplete-field.tsx`, while component is named `AutocompleteSelect`. Why is there this field/select naming difference?

Go back to my last review where I requested changes and make sure everything is addressed.

unocha-hpc reviewed (approved):
Checks have passed and this pull request is ready for manual review

Onitoxan committed:
Replace xlxs package with exceljs

Onitoxan committed:
Abstract function to avoid repetitive code

unocha-hpc reviewed (approved):
Checks have passed and this pull request is ready for manual review

Onitoxan commented:
@Pl217  Regarding I think the last comment should be fully addressed. Just to also answer to why the file name is `autocomplete-field.tsx` and the component `AutcompleteSelect`, the name of the file should be `autocomplete-select-field.tsx` is just that at the beginning when I was creating this component, I didn't have clear how to name the files. If you don't mind I would prefer to change this at the end of the development (Or at least in a more recent branch) so rebasing doesn't overcomplicate

unocha-hpc reviewed (approved):
Checks have passed and this pull request is ready for manual review

Pl217 reviewed (changes requested):
Unfortunately, there are still comments which aren't addressed nor commented back on. Look at all inline comments which aren't resolved.

> Regarding I think the last comment should be fully addressed. Just to also answer to why the file name is `autocomplete-field.tsx` and the component `AutcompleteSelect`, the name of the file should be `autocomplete-select-field.tsx` is just that at the beginning when I was creating this component, I didn't have clear how to name the files. If you don't mind I would prefer to change this at the end of the development (Or at least in a more recent branch) so rebasing doesn't overcomplicate

I don't mind if we rename later, but I am afraid we'll forget, because PRs stay really long in code review.

Onitoxan committed:
Remove irrelevant comment

Onitoxan committed:
Rewrite handleAbortController

Onitoxan committed:
Remove unused props

unocha-hpc reviewed (approved):
Checks have passed and this pull request is ready for manual review

Onitoxan commented:
@Pl217 I wrote it down in my notes to not forget, anyways I'll just doing in my current working branch so is not forgotten

> I don't mind if we rename later, but I am afraid we'll forget, because PRs stay really long in code review.

Pl217 reviewed (approved):
This work is finally approved, but I want to comment on something I caught in your most recent inline comment, where you expressed your thoughts about the review process:

> For this kind of PR I think we should be a bit more flexible since is not finished work, once I finish I will review myself everything and in that moment I think is nice if you are very picky so everything can be addressed because it has to be.

I have seen so much dead code across our codebases, where the code was unused from the beginning. Such dead code was added as part of a much bigger change and it slipped in. When that code becomes legacy (which could be argued is immediately after merging), other people reading it later give it some meaning, thinking it does something and being afraid to remove it, since it is legacy code.

When it comes to unused properties, they are easy to spot later, but there could be some more complex leftovers from the development process which no one will have the courage to touch later when it becomes legacy. And, given that you've written some parts of this code more than a year ago, the code is foreign to you too.

Plus, if you make big refactorings later, like you expressed that you want to rename the components once you have clearer names (and purposes) in mind, that will produce git history that is hard to track.

unocha-hpc reviewed (approved):
Checks have passed and this pull request is ready for manual review
